libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
+            sorted.reserve(std::ranges::size(r));
+         }
+
+         size_t index = 0;
+         for (const auto& entry : r) {
+            sorted.emplace_back(entry, index);
+            ++index;
+         }
+
+         std::ranges::stable_sort(sorted, [](auto&& lhs, auto&& rhs) {
+            return lhs.first < rhs.first;
+         });
+
+         return sorted;
+      };
+
+      auto orig_sorted = sort_input(orig);
+      auto targ_sorted = sort_input(target);
+
+      size_t orig_index = 0;
+      size_t cur_index  = 0;
+
+      while ((orig_index < orig_sorted.size()) || (cur_index < targ_sorted.size())) {
+         std::weak_ordering cmp = std::weak_ordering::equivalent;
+
+         if (orig_sorted.size() <= orig_index) {
+            cmp = std::weak_ordering::greater;
+         } else if (targ_sorted.size() <= cur_index) {
+            cmp = std::weak_ordering::less;
+         } else {
+            cmp = (orig_sorted[orig_index].first <=> targ_sorted[cur_index].first);
+         }
+
+         if (cmp == std::weak_ordering::less) {
+            assert(orig_sorted[orig_index].second <= std::numeric_limits<SizeType>::max());
+            result.remove_indexes.emplace_back(orig_sorted[orig_index].second);
+            ++orig_index;
+         } else if (cmp == std::weak_ordering::greater) {
+            assert(targ_sorted[cur_index].second <= std::numeric_limits<SizeType>::max());
+            result.insert_indexes.emplace_back(targ_sorted[cur_index].second, std::move(targ_sorted[cur_index].first));
+            ++cur_index;
+         } else {
+            // cmp == std::weak_ordering::equivalent
+            assert(orig_sorted[orig_index].first == targ_sorted[cur_index].first);
+            if (orig_sorted[orig_index].second != targ_sorted[cur_index].second) {
+               assert(orig_sorted[orig_index].second <= std::numeric_limits<SizeType>::max());
+               assert(targ_sorted[cur_index].second <= std::numeric_limits<SizeType>::max());
+               result.move_indexes.emplace_back(orig_sorted[orig_index].second, targ_sorted[cur_index].second);
             }
-         } else if (s < source.size()) {
-            // remove extra in source
-            assert(s <= std::numeric_limits<SizeType>::max());
-            result.remove_indexes.push_back(s);
-            ++s;
-         } else if (t < target.size()) {
-            // insert extra in target
-            assert(t <= std::numeric_limits<SizeType>::max());
-            result.insert_indexes.emplace_back(t, target[t]);
-            ++t;
+            ++orig_index;
+            ++cur_index;
          }
       }
 
+      std::ranges::sort(result.remove_indexes, [](auto&& lhs, auto&& rhs) { return lhs > rhs; }); // descending
+      std::ranges::sort(result.insert_indexes, [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+      std::ranges::sort(result.move_indexes,   [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+
+      // update move indexes
```

Comment:
This section of updating move indexes is quite complicated. Can it somehow be avoided by tweaking other parts of the algorithm?